core/src/main/scala/org/apache/spark/TaskEndReason.scala

```diff
@@ -31,14 +31,18 @@ import org.apache.spark.util.Utils
  * old stages to be resubmitted, such as shuffle map fetch failures.
  */
 @DeveloperApi
-sealed trait TaskEndReason
+sealed trait TaskEndReason {
+  def shouldTaskEndEventuallyFailJob: Boolean
```

Comment:
Methods in the scheduler code only get instances of TaskEndReason, not specifically TaskFailedReason. Although I suppose we could do case matching / casting to appropriately handle where it's important.